enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,689 @@
+
+```
+[root@ovn-worker ~]# ovs-ofctl show breth0
+OFPT_FEATURES_REPLY (xid=0x2): dpid:00000242ac120003
+n_tables:254, n_buffers:0
+capabilities: FLOW_STATS TABLE_STATS PORT_STATS QUEUE_STATS ARP_MATCH_IP
+actions: output enqueue set_vlan_vid set_vlan_pcp strip_vlan mod_dl_src mod_dl_dst mod_nw_src mod_nw_dst mod_nw_tos mod_tp_src mod_tp_dst
+ 1(eth0): addr:02:42:ac:12:00:03
+     config:     0
+     state:      0
+     current:    10GB-FD COPPER
+     speed: 10000 Mbps now, 0 Mbps max
+ 2(patch-breth0_ov): addr:ca:2c:82:e9:06:42
+     config:     0
+     state:      0
+     speed: 0 Mbps now, 0 Mbps max
+
+
+[root@ovn-worker ~]# ovs-ofctl dump-flows breth0 |grep 10.244
+ cookie=0xdeff105, duration=437.393s, table=0, n_packets=4, n_bytes=392, idle_age=228, priority=109,ip,in_port=2,dl_src=02:42:ac:12:00:03,nw_src=10.244.1.0/24 actions=ct(commit,zone=64000,exec(load:0x1->NXM_NX_CT_MARK[])),output:1
+ cookie=0xdeff105, duration=437.393s, table=0, n_packets=0, n_bytes=0, idle_age=437, priority=104,ip,in_port=2,nw_src=10.244.0.0/16 actions=drop
+ cookie=0xdeff105, duration=437.393s, table=1, n_packets=4, n_bytes=392, idle_age=228, priority=15,ip,nw_dst=10.244.0.0/16 actions=output:2
+
+```
+
+Packets matching the pod IP are forwarded directly to OVN, where they are forwarded to the pod without SNAT. Additional
+flows will need to be created for additional networks. For shared gateway mode, the reply packet will always return via
+OVS and follow a symmetrical path. This is also true if local gateway mode is being used in combination with Multiple
+External Gateways (MEG). However, if local gateway mode is used without MEG, then the reply packet will be forwarded into
+the kernel networking stack, where it will routed out br-ex via the kernel routing table.
+
+###### Ingress via a Secondary NIC
+
+If packets enter into the host via a NIC that is not attached to OVS, the kernel networking stack will forward it into
+the proper VRF, where it will be forwarded via ovn-k8s-mp0 of the respective user-defined network. Today when these
+packets ingress ovn-k8s-mp0 into OVN, they are SNAT'ed to the ovn-k8s-mp0 address (.2 address). Reasons for this include:
+1. To ensure incoming traffic that may be routed to another node via geneve would return back to this node. This is undesirable
+and unnecessary for networks where only their respective per-node pod subnet is being advertised to the BGP fabric.
+2. If running shared gateway mode, the reply packet would be routed via br-ex with the default gateway configured in
+ovn_cluster_router.
+
+For local gateway mode, changes will need to be made to skip the SNAT as it is not necessary and provides pods with the
+true source IP of the sender.
+
+
+Additionally, modifications will need to be made in the kernel routing table to
+leak routes to the pod subnets from each user-defined VRF into the default VRF routing table.
+
+#### Deployment Considerations
```

Comment:
Can we refine the deployment workflow and the CNO api here @trozet @jcaamano ?

From my point of view, there are two operating modes:
- the user needs bgp from the start and in ovnk
- the user doesn't care about bgp on ovnk but wants metallb

In the former case, when the metallb-operator needs to deploy, it will find the cno flag already set and do nothing.

In the latter case, the metallb-operator will need to set the CNO flag to tell CNO to deploy frr-k8s.

Reply:
i've updated the text some, and included the API. We need to discuss how upgrade will work from 4.16 ->4.17.

Reply:
we talked offline and decided on an upgrade strategy. will update the doc